silo-core/contracts/incentives/base/DistributionManager.sol

```diff
@@ -0,0 +1,318 @@
+import {EnumerableSet} from "openzeppelin5/utils/structs/EnumerableSet.sol";
+
+import {IDistributionManager} from "../interfaces/IDistributionManager.sol";
+import {DistributionTypes} from "../lib/DistributionTypes.sol";
+import {TokenHelper} from "../../lib/TokenHelper.sol";
+
+/**
+ * @title DistributionManager
+ * @notice Accounting contract to manage multiple staking distributions
+ */
+contract DistributionManager is IDistributionManager, Ownable2Step {
+    using EnumerableSet for EnumerableSet.Bytes32Set;
+
+    EnumerableSet.Bytes32Set internal _incentivesProgramIds;
+
+    mapping(bytes32 => IncentivesProgram) public incentivesPrograms;
+
+    address public immutable NOTIFIER; // solhint-disable-line var-name-mixedcase
+
+    uint8 public constant PRECISION = 18;
+    uint256 public constant TEN_POW_PRECISION = 10 ** PRECISION;
+
+    modifier onlyNotifier() {
+        if (msg.sender != NOTIFIER) revert OnlyNotifier();
+        _;
+    }
+
+    constructor(address _owner, address _notifier) Ownable(_owner) {
+        NOTIFIER = _notifier;
+    }
+
+    /// @inheritdoc IDistributionManager
+    function setDistributionEnd(
+        string calldata _incentivesProgram,
+        uint40 _distributionEnd
+    ) external onlyOwner {
+        bytes32 incentivesProgramId = getProgramId(_incentivesProgram);
+        incentivesPrograms[incentivesProgramId].distributionEnd = _distributionEnd;
+
+        emit DistributionEndUpdated(_incentivesProgram, _distributionEnd);
+    }
+
+    /// @inheritdoc IDistributionManager
+    function getDistributionEnd(string calldata _incentivesProgram) external view override returns (uint256) {
+        bytes32 incentivesProgramId = getProgramId(_incentivesProgram);
+        return incentivesPrograms[incentivesProgramId].distributionEnd;
+    }
+
+    /// @inheritdoc IDistributionManager
+    function getUserData(address _user, string calldata _incentivesProgram) public view override returns (uint256) {
+        bytes32 incentivesProgramId = getProgramId(_incentivesProgram);
+        return incentivesPrograms[incentivesProgramId].users[_user];
+    }
+
+    /// @inheritdoc IDistributionManager
+    function getIncentivesProgramData(string calldata _incentivesProgram)
+        public
+        view
+        override
+        returns (
+            uint256 index,
+            uint256 emissionPerSecond,
+            uint256 lastUpdateTimestamp,
+            uint256 distributionEnd
+        )
+    {
+        bytes32 incentivesProgramId = getProgramId(_incentivesProgram);
+
+        index = incentivesPrograms[incentivesProgramId].index;
+        emissionPerSecond = incentivesPrograms[incentivesProgramId].emissionPerSecond;
+        lastUpdateTimestamp = incentivesPrograms[incentivesProgramId].lastUpdateTimestamp;
+        distributionEnd = incentivesPrograms[incentivesProgramId].distributionEnd;
+    }
+
+    /// @inheritdoc IDistributionManager
+    function incentivesProgram(string calldata _incentivesProgram)
```

Comment:
Why do we need both `getIncentivesProgramData` and `incentivesProgram`?

Reply:
Done. Removed `getIncentivesProgramData()` fn.